### ISSUE TITLE
Wrong result in malware samples comparison

### ISSUE THREAD

**Issue by im-overlord04** [sdhash/sdhash#17]
I'm analyzing some malware samples to check their similarity. I met two samples giving 100 similarity but at the same time they are totally different. By using 010 Editor I found that the only bytes they share are the PE MS-DOS header and some sequences of 0s, so their similarity should be around 0. You can find those two samples in this zip archive: [samples.zip](https://github.com/sdhash/sdhash/files/10133705/samples.zip).
THOSE ARE REAL WINDOWS MALWARE, DON'T EXECUTE THEM UNLESS YOU ARE IN A CONTROLLED ENVIRONMENT. The password of the archive is: "**infected**". What I'd like to understand is if this is a bug in sdhash algorithm, sdhash implementation or if this is the expected behavior.

**Comment by utkonos** [comment]
@im-overlord04 

This is a bug in the stream mode scoring function, not in the algorithm itself, and DD mode handles it correctly as the following output shows:

```text
=== comparisons ===
stream  vs  stream  100
dd      vs  dd      0
stream  vs  dd      2
dd      vs  stream  100
```

The root cause is low feature density. sdhash works by extracting unique features from the input using a sliding entropy window. When a file consists largely of repeated byte sequences, such as long runs of zeros, repetitive headers, padding, the deduplication filters reject almost everything and very few features are inserted into the bloom filters. In the extreme case, both files produce a single bloom filter with fewer than 16 elements each.

The stream mode scoring function has a known sparse filter threshold: any filter with fewer than 16 elements is considered too sparse to score reliably and is skipped. When there is only one filter and it falls below this threshold, the sparse filter exclusion logic does not apply because excluding the only filter would leave nothing to score. The function divides by 1 and scores the two nearly-empty bloom filters against each other. Two nearly-empty filters produce an artificially high match because there is so little signal that the scoring math breaks down.

DD mode produces the correct answer of 0 because it compares the files block by block and the blocks genuinely share no features. For files with low entropy or heavy repetition, DD mode should be preferred and stream mode scores should be treated with caution or discarded entirely.

The correct fix is to detect degenerate digests before scoring and return a sentinel value rather than a misleading score. The detection criterion is feature density: total unique features inserted divided by input size. A normal high-entropy binary produces consistent density across its length. A zero-padded or repetitive file produces a density so low that the resulting digest does not contain enough information to support a meaningful similarity comparison.

I am implementing a solution in a Go port of this library. I will document the fix thoroughly in this issue once the implementation is complete, so the fix can be backported to the reference implementation if the authors wish to take it up.

The Go implementation is found here:

https://github.com/malwarology/sdhash

**Comment by utkonos** [comment]
This issue is related to another issue. Both have the same underlying root cause.

https://github.com/sdhash/sdhash/issues/5

**Comment by utkonos** [comment]
Both samples produce degenerate digests: too few features for a meaningful comparison. Our Go port now exposes `FeatureDensity()` (total unique features / input size) so callers can detect this before trusting the score:

```
=== 0091e9304703d6a665e2aa2b399262919f7376555e1a8c583523d404e616d64b ===
  stream   filters=6    input=100850     density=0.008428  [DEGENERATE]
  dd       filters=2    input=100850     density=0.002241  [DEGENERATE]
=== 03b576e49e9162e768524b4a0d129ea8c48074f527c53b1c48a2fb9b0213e284 ===
  stream   filters=94   input=1212896    density=0.012314  [DEGENERATE]
  dd       filters=19   input=1212896    density=0.002919  [DEGENERATE]
=== comparisons ===
  stream  vs  stream  100
  dd      vs  dd      0
  stream  vs  dd      2
  dd      vs  stream  100
```

The stream score of 100 is a false positive. Both densities fall well below 0.02; checking density before comparing would have caught this.

Implementation: [malwarology/sdhash@fcd6fb2](https://github.com/malwarology/sdhash/commit/fcd6fb2a1bdc52691bf97b809e194f0f1ed1e0d0)

Side note: this issue has the same underlying root cause as Issue #5, it just comes at the issue from the opposite direction.
